src/modules/0.0.9/aws_vpc/index.ts

```diff
@@ -278,14 +289,38 @@ export const AwsVpcModule: Module2 = new Module2({
             return out;
           }
         },
+        updateOrReplace: (a: NatGateway, b: NatGateway) => {
+          if (!(Object.is(a.state, b.state) && AwsVpcModule.utils.eqTags(a.tags, b.tags))
+            && Object.is(a.connectivityType, b.connectivityType)
+            && Object.is(a.elasticIp?.allocationId, b.elasticIp?.allocationId)
+            && Object.is(a.subnet?.subnetId, b.subnet?.subnetId)) return 'update';
+          return 'replace';
+        },
         update: async (es: NatGateway[], ctx: Context) => {
-          // There is no update mechanism for a NatGateway so instead we will restore the values
+          const client = await ctx.getAwsClient() as AWS;
           const out = [];
           for (const e of es) {
             const cloudRecord = ctx?.memo?.cloud?.NatGateway?.[e.natGatewayId ?? ''];
-            cloudRecord.id = e.id;
-            await AwsVpcModule.mappers.natGateway.db.update(cloudRecord, ctx);
-            out.push(cloudRecord);
+            const isUpdate = Object.is(AwsVpcModule.mappers.natGateway.cloud.updateOrReplace(cloudRecord, e), 'update');
+            if (isUpdate) {
+              if (!AwsVpcModule.utils.eqTags(cloudRecord.tags, e.tags)) {
```

Comment:
Because `update` and "`restore`" are mixed in the `isUpdate` path. Here we have these 3 options:
- `isUpdate` and `tags` are modified means we need to update the tags and once we have the updated object that will `restore` if the `state` property changed.
- `isUpdate` and just `state` was modified. This means we just need to restore the record. 
- !`isUpdate` == `replace` Replace path

Reply:
Yes, but I'm saying that `isUpdate` is the outer conditional, so it's `true` for this path, so why is `updatedNatGateway` within it? I understand it being on line 316, but not line 311, basically.

Reply:
because for the restore path I do not want to re-query the cloud, just using the cloudRecord from memo to restore. I'm finishing each branch with a `continue`. Maybe it will be easier to understand if I do an `if - else if - else` branches with some comments? 
- `if` being the tags update
- `else if` the restore path
- `else` the restore

Reply:
Wait, why two "restore" paths, now? I thought one update was tags, one update was state, one replace is every other property? What causes a "restore"?

Reply:
what do you mean? is only one restore path, for the state.

Reply:
![Screenshot from 2022-06-17 11-37-49](https://user-images.githubusercontent.com/765551/174340415-d4375e6b-df20-4757-83fd-b8587d48ec74.png)

Reply:
@aguillenv I am saying that you call the `else if` path "the restore path" and you call the `else` path "the restore" so there are two restores according to you?

Reply:
ohh typo, my bad. else would be the `replace`

Reply:
```suggestion
            // `isUpdate` means only `tags` and/or `state` have changed
            if (isUpdate && !AwsVpcModule.utils.eqTags(cloudRecord.tags, e.tags)) {
              // If tags have change, no matter is state change or nor, we update the tags, call AWS and update
              await client.updateTags(e.natGatewayId ?? '', e.tags);
              const rawNatGateway = await client.getNatGateway(e.natGatewayId ?? '');
              const updatedNatGateway = await AwsVpcModule.utils.natGatewayMapper(rawNatGateway, ctx);
              updatedNatGateway.id = e.id;
              await AwsVpcModule.mappers.natGateway.db.update(updatedNatGateway, ctx);
              out.push(updatedNatGateway);
            } else if (isUpdate && AwsVpcModule.utils.eqTags(cloudRecord.tags, e.tags)) {
              // If tags have **not** change, it means only state changed. This is the restore path. We do not call AWS again, just use the record we have in memo.
              cloudRecord.id = e.id;
              await AwsVpcModule.mappers.natGateway.db.update(cloudRecord, ctx);
              out.push(cloudRecord);
            } else {
              // Replace path
              // Need to delete first to make the elastic ip address available
              await AwsVpcModule.mappers.natGateway.cloud.delete(cloudRecord, ctx);
              const newNatGateway = await AwsVpcModule.mappers.natGateway.cloud.create(e, ctx);
              out.push(newNatGateway);
            }
```

Reply:
Okay, now on my original question: why is the NAT gateway touched when tags are changed?

Reply:
Wait,  I just remembered that we decided against a centralized tag system and have tags on the entities directly. Nevermind about this question, but I do like the clearer code.

Reply:
I was like 
![Alt Text](https://media.giphy.com/media/3owzW5c1tPq63MPmWk/giphy.gif)